File: clients/android/app/src/main/java/network/buildit/modules/training/integration/TrainingCallingIntegration.kt
package network.buildit.modules.training.integration

import android.util.Log
import kotlinx.coroutines.flow.first
import network.buildit.modules.training.domain.model.*
import network.buildit.modules.training.domain.repository.TrainingRepository
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Conference room configuration for training sessions.
 */
data class TrainingConferenceConfig(
    val name: String,
    val maxParticipants: Int? = null,
    val waitingRoom: Boolean = true,
    val allowRecording: Boolean = true,
    val e2eeRequired: Boolean = true,
    val instructorPubkey: String
)

/**
 * Conference room result.
 */
data class ConferenceRoom(
    val id: String,
    val joinUrl: String,
    val hostUrl: String?
)

/**
 * Integration between Training module and Calling module.
 * Provides live training sessions via video conferencing.
 */
@Singleton
class TrainingCallingIntegration @Inject constructor(
    private val trainingRepository: TrainingRepository
    // In a real implementation, would inject CallingManager
) {
    companion object {
        private const val TAG = "TrainingCallingIntegration"
    }

    /**
     * Creates a conference room for a live training session.
     *
     * @param lesson The live session lesson
     * @param config Conference configuration
     * @return Conference room details
     */
    suspend fun createLiveSession(
        lesson: Lesson,
        config: TrainingConferenceConfig
    ): ConferenceRoom {
        require(lesson.type == LessonType.LIVE_SESSION) {
            "Lesson must be a live session"
        }

        val content = lesson.content as LessonContent.LiveSession

        // Generate conference room ID
        val conferenceRoomId = "training-${lesson.id}-${UUID.randomUUID().toString().take(8)}"

        Log.i(TAG, "Creating conference room for training session: $conferenceRoomId")

        // In a real implementation, this would:
        // 1. Call CallingManager to create a conference room
        // 2. Configure room settings (E2EE, recording, etc.)
        // 3. Set up instructor as host

        // Simulated conference room creation
        val joinUrl = "/conference/$conferenceRoomId"
        val hostUrl = "/conference/$conferenceRoomId/host"

        // Update the lesson with the conference room ID
        val updatedContent = content.copy(conferenceRoomId = conferenceRoomId)
        val updatedLesson = lesson.copy(
            content = updatedContent,
            updated = System.currentTimeMillis() / 1000
        )
        trainingRepository.updateLesson(updatedLesson)

        return ConferenceRoom(
            id = conferenceRoomId,
            joinUrl = joinUrl,
            hostUrl = hostUrl
        )
    }

    /**
     * Starts a live training session.
     * Sends join notifications to RSVPed participants.
     *
     * @param lessonId The lesson ID
     */
    suspend fun startLiveSession(lessonId: String) {
        val lesson = trainingRepository.getLesson(lessonId).first()
            ?: throw IllegalArgumentException("Lesson not found")

        require(lesson.type == LessonType.LIVE_SESSION) {
            "Lesson must be a live session"
        }

        val content = lesson.content as LessonContent.LiveSession

        requireNotNull(content.conferenceRoomId) {
            "Conference room not created"
        }

        Log.i(TAG, "Starting live training session: $lessonId")

        // In a real implementation, this would:
        // 1. Get all RSVPed users
        // 2. Send push notifications with join link
        // 3. Start the conference room
        // 4. Enable recording if configured

        val rsvps = trainingRepository.getLiveSessionRSVPs(lessonId).first()
        val confirmedRsvps = rsvps.filter { it.status == RSVPStatus.CONFIRMED }

        Log.i(TAG, "Notifying ${confirmedRsvps.size} confirmed participants")

        // Would send notifications via MessagingManager or NotificationService
    }

    /**
     * Ends a live training session and saves the recording.
     *
     * @param lessonId The lesson ID
     * @return Recording URL if available
     */
    suspend fun endLiveSession(lessonId: String): String? {
        val lesson = trainingRepository.getLesson(lessonId).first()
            ?: throw IllegalArgumentException("Lesson not found")

        require(lesson.type == LessonType.LIVE_SESSION) {
            "Lesson must be a live session"
        }

        val content = lesson.content as LessonContent.LiveSession

        Log.i(TAG, "Ending live training session: $lessonId")

        // In a real implementation, this would:
        // 1. End the conference room
        // 2. Get the recording URL
        // 3. Process and store the recording

        val recordingUrl = content.recordingUrl // Would come from calling module

        // Update lesson with recording URL if available
        if (recordingUrl != null) {
            val updatedContent = content.copy(recordingUrl = recordingUrl)
            val updatedLesson = lesson.copy(
                content = updatedContent,
                updated = System.currentTimeMillis() / 1000
            )
            trainingRepository.updateLesson(updatedLesson)
        }

        return recordingUrl
    }

    /**
     * Tracks attendance for a live training session.
     *
     * @param lessonId The lesson ID
     * @param pubkey The participant's public key
     * @param duration Duration attended in seconds
     */
    suspend fun trackLiveAttendance(
        lessonId: String,
        pubkey: String,
        duration: Long
    ) {
        val lesson = trainingRepository.getLesson(lessonId).first()
            ?: throw IllegalArgumentException("Lesson not found")

        require(lesson.type == LessonType.LIVE_SESSION) {
            "Lesson must be a live session"
        }

        val content = lesson.content as LessonContent.LiveSession
        val now = System.currentTimeMillis() / 1000

        // Calculate if this was a complete session attendance
        val sessionDurationSeconds = content.duration * 60L
        val wasCompleteSession = duration >= (sessionDurationSeconds * 0.8) // 80% threshold

        val attendance = LiveSessionAttendance(
            id = UUID.randomUUID().toString(),
            lessonId = lessonId,
            pubkey = pubkey,
            joinedAt = now - duration,
            leftAt = now,
            duration = duration,
            wasCompleteSession = wasCompleteSession
        )

        trainingRepository.recordLiveSessionAttendance(attendance)

        Log.i(TAG, "Recorded attendance: $pubkey attended ${duration / 60} minutes (complete: $wasCompleteSession)")

        // If attended for significant duration, mark lesson progress
        if (duration > 30 * 60) { // 30 minutes minimum
            val existingProgress = trainingRepository.getLessonProgress(lessonId, pubkey).first()
            if (existingProgress == null || existingProgress.status != ProgressStatus.COMPLETED) {
                val progress = LessonProgress(
                    id = existingProgress?.id ?: UUID.randomUUID().toString(),
                    lessonId = lessonId,
                    pubkey = pubkey,
                    status = if (wasCompleteSession) ProgressStatus.COMPLETED else ProgressStatus.IN_PROGRESS,
                    score = null,
                    timeSpent = duration,
                    lastPosition = null,
                    completedAt = if (wasCompleteSession) now else null,
                    attempts = null,
                    created = existingProgress?.created ?: now,
                    updated = now
                )
                trainingRepository.saveProgress(progress)
            }
        }
    }

    /**
     * Gets conference room details for a training session.
     *
     * @param lessonId The lesson ID
     * @return Conference room status
     */
    suspend fun getConferenceDetails(lessonId: String): ConferenceStatus {
        val lesson = trainingRepository.getLesson(lessonId).first()
            ?: return ConferenceStatus(null, false, 0)

        if (lesson.type != LessonType.LIVE_SESSION) {
            return ConferenceStatus(null, false, 0)
        }

        val content = lesson.content as LessonContent.LiveSession

        // In a real implementation, would query the calling module for room status
        return ConferenceStatus(
            conferenceRoomId = content.conferenceRoomId,
            isActive = false, // Would check with calling module
            participantCount = 0 // Would get from calling module
        )
    }

    /**
     * Checks if the calling module is available and properly configured.
     */
    fun isCallingModuleAvailable(): Boolean {
        // In a real implementation, would check:
        // 1. If calling module is enabled
        // 2. If WebRTC is properly initialized
        // 3. If user has necessary permissions
        return true // Placeholder
    }
}

/**
 * Conference status information.
 */
data class ConferenceStatus(
    val conferenceRoomId: String?,
    val isActive: Boolean,
    val participantCount: Int
)

File: clients/ios/BuildIt/Modules/Training/Integrations/TrainingCallingIntegration.swift
// TrainingCallingIntegration.swift
// BuildIt - Decentralized Mesh Communication
//
// Integration between Training and Calling modules for live training sessions.

import Foundation
import os.log

/// Conference room configuration for training
public struct TrainingConferenceConfig: Sendable {
    public var name: String
    public var maxParticipants: Int?
    public var waitingRoom: Bool
    public var allowRecording: Bool
    public var e2eeRequired: Bool
    public var instructorPubkey: String

    public init(
        name: String,
        maxParticipants: Int? = nil,
        waitingRoom: Bool = true,
        allowRecording: Bool = true,
        e2eeRequired: Bool = true,
        instructorPubkey: String
    ) {
        self.name = name
        self.maxParticipants = maxParticipants
        self.waitingRoom = waitingRoom
        self.allowRecording = allowRecording
        self.e2eeRequired = e2eeRequired
        self.instructorPubkey = instructorPubkey
    }
}

/// Result of creating a live session
public struct LiveSessionCreationResult: Sendable {
    public var conferenceRoomId: String
    public var joinUrl: String

    public init(conferenceRoomId: String, joinUrl: String) {
        self.conferenceRoomId = conferenceRoomId
        self.joinUrl = joinUrl
    }
}

/// Conference room details
public struct ConferenceDetails: Sendable {
    public var conferenceRoomId: String?
    public var isActive: Bool
    public var participantCount: Int

    public init(conferenceRoomId: String? = nil, isActive: Bool = false, participantCount: Int = 0) {
        self.conferenceRoomId = conferenceRoomId
        self.isActive = isActive
        self.participantCount = participantCount
    }
}

/// Integration between Training and Calling modules
@MainActor
public class TrainingCallingIntegration {
    // MARK: - Properties

    private let manager: TrainingManager
    private let logger = Logger(subsystem: "com.buildit", category: "TrainingCallingIntegration")

    // MARK: - Initialization

    public init(manager: TrainingManager) {
        self.manager = manager
    }

    // MARK: - Live Session Management

    /// Create a conference room for a live training session
    public func createLiveSession(
        lesson: Lesson,
        config: TrainingConferenceConfig
    ) async throws -> LiveSessionCreationResult {
        guard lesson.type == .liveSession else {
            throw TrainingError.invalidData
        }

        // Generate conference room ID
        let conferenceRoomId = "training-\(lesson.id)-\(UUID().uuidString.prefix(8))"

        // In a real implementation, this would call the CallingModule to create a room
        logger.info("Creating conference room for training session: \(conferenceRoomId)")

        // Generate join URL
        let joinUrl = "/conference/\(conferenceRoomId)"

        // Update the lesson with conference room ID
        if case .liveSession(var liveContent) = lesson.content {
            liveContent.conferenceRoomId = conferenceRoomId
            logger.info("Created training conference room: \(conferenceRoomId)")
        }

        return LiveSessionCreationResult(
            conferenceRoomId: conferenceRoomId,
            joinUrl: joinUrl
        )
    }

    /// Start a live training session
    /// Sends join links to enrolled/RSVPed users
    public func startLiveSession(lessonId: String) async throws {
        guard let lesson = try await manager.getLesson(id: lessonId),
              lesson.type == .liveSession else {
            throw TrainingError.lessonNotFound
        }

        guard case .liveSession(let liveContent) = lesson.content,
              let _ = liveContent.conferenceRoomId else {
            throw TrainingError.invalidData
        }

        // In a real implementation, this would:
        // 1. Get all RSVPed users
        // 2. Send notifications with join links
        // 3. Start the conference room

        logger.info("Starting live training session: \(lessonId)")
    }

    /// End a live training session and save recording
    public func endLiveSession(lessonId: String) async throws -> String? {
        guard let lesson = try await manager.getLesson(id: lessonId),
              lesson.type == .liveSession else {
            throw TrainingError.lessonNotFound
        }

        guard case .liveSession(let liveContent) = lesson.content else {
            throw TrainingError.invalidData
        }

        // In a real implementation, this would:
        // 1. End the conference
        // 2. Get the recording URL
        // 3. Process and store the recording

        let recordingUrl = liveContent.recordingUrl

        logger.info("Ended live training session: \(lessonId), recording: \(recordingUrl ?? "none")")

        return recordingUrl
    }

    /// Track attendance for a live training session
    public func trackLiveAttendance(
        lessonId: String,
        pubkey: String,
        joinedAt: Date,
        leftAt: Date? = nil
    ) async throws {
        try await manager.recordLiveAttendance(
            lessonId: lessonId,
            joinedAt: joinedAt,
            leftAt: leftAt
        )

        let duration = (leftAt ?? Date()).timeIntervalSince(joinedAt)
        let durationMinutes = Int(duration / 60)

        logger.info("Recorded training attendance: \(pubkey) attended \(durationMinutes) minutes")

        // If attended for significant duration, mark progress
        if duration > 30 * 60 { // 30 minutes
            try await manager.completeLesson(lessonId)
        }
    }

    /// Get conference room details for a training session
    public func getConferenceDetails(lessonId: String) async throws -> ConferenceDetails {
        guard let lesson = try await manager.getLesson(id: lessonId),
              lesson.type == .liveSession else {
            return ConferenceDetails()
        }

        guard case .liveSession(let liveContent) = lesson.content else {
            return ConferenceDetails()
        }

        // In a real implementation, would query the CallingModule for room status
        return ConferenceDetails(
            conferenceRoomId: liveContent.conferenceRoomId,
            isActive: false, // Would check with CallingModule
            participantCount: 0 // Would get from CallingModule
        )
    }

    /// Check if CallingModule is available
    public func isCallingModuleAvailable() async -> Bool {
        // In a real implementation, would check if CallingModule is enabled
        // and properly configured
        return true
    }

    /// Create a group call for a training session
    public func createGroupCall(
        lesson: Lesson,
        invitedPubkeys: [String]
    ) async throws -> String {
        guard lesson.type == .liveSession else {
            throw TrainingError.invalidData
        }

        // In a real implementation, this would call CallingModule.createGroupCall()
        let roomId = "training-call-\(lesson.id)"

        logger.info("Created group call for training: \(roomId) with \(invitedPubkeys.count) invitees")

        return roomId
    }

    /// Join a training session call
    public func joinTrainingCall(roomId: String) async throws {
        // In a real implementation, would call CallingModule.joinGroupCall(roomId:)
        logger.info("Joining training call: \(roomId)")
    }

    /// Leave a training session call
    public func leaveTrainingCall(roomId: String) async throws {
        // In a real implementation, would call CallingModule.leaveGroupCall(roomId:)
        logger.info("Leaving training call: \(roomId)")
    }
}
